### ISSUE TITLE
Backend rate limit tests are failing

### ISSUE THREAD

**Issue by to-sta** [activist-org/activist#879]
### Terms

- [X] I have searched all [open bug reports](https://github.com/activist-org/activist/issues?q=is%3Aopen+is%3Aissue+label%3Abug)
- [X] I agree to follow activist's [Code of Conduct](https://github.com/activist-org/activist/blob/main/.github/CODE_OF_CONDUCT.md)

### Behavior

Our rate limit tests are currently failing, but the rate limiting functionality is working correctly. I have manually tested it using SwaggerUI.

**Comment by to-sta** [comment]
cc @josueemartinezz

**Comment by thesarfo** [comment]
is this being worked on? If not I could get this done

**Comment by to-sta** [comment]
Would be great if you can tackle this @thesarfo

**Comment by andrewtavis** [comment]
Thanks for the assignment, @to-sta, and thanks for offering to help out @thesarfo :)

**Comment by to-sta** [comment]
@prajguru54 thanks for checking :). I removed the tests for a while in order to avoid failing gh actions. Forgot to add a comment here. The last change to the test was made in the commit ` 591bd07c243c279573cbf5b9eeb51e4ca5be89b0`. But I think it would be better to start from scratch 😄. 

We want to check the rate limits for signed in users and anom users.

**Comment by andrewtavis** [comment]
@prajguru54 or @thesarfo, would either of you want to reimplement them?

**Comment by thesarfo** [comment]
yes, I would want to do that. but @prajguru54 can take it if he feels up to it

**Comment by andrewtavis** [comment]
You're assigned, @thesarfo, so by all means feel free to send something along :) Let us know if you'd like some ideas or support!

**Comment by prajguru54** [comment]
@andrewtavis , I will be glad to take this. Already started on it yesterday. @thesarfo  Have you also started as well??

**Comment by andrewtavis** [comment]
Hey all 👋 Checking in post a busy last month on my end. Hope all are well :) We're still looking for the tests that were removed to be re-implemented. @thesarfo or @prajguru54: Do either of you have interest?

**Comment by thesarfo** [comment]
@andrewtavis I’ve been going through the codebase to locate the rate-limiting functionality but haven’t been able to pin it down yet. I’m keen to write the tests for it, could you point me in the right direction?

**Comment by andrewtavis** [comment]
CC @to-sta who might have a bit better of an opinion on how to implement this? :)

**Comment by to-sta** [comment]
Hi @thesarfo, the functionality comes from DRF, here are the [docs](https://www.django-rest-framework.org/api-guide/throttling/). 

We have set default values users and anons in settings.py. 

For your test you could also overwrite those settings and test against that 🙂.

You can test the rate limiting yourself via SwaggerUI. For that you need to run the backend container and go to http://localhost:8000/v1/schema/swagger-ui/

**Comment by andrewtavis** [comment]
Quick check in here, @thesarfo 👋 Do you still plan on working on this issue? :)

Hope all's well!

Given the timing here, also @prajguru54 please let us know if this would be of interest 😊

**Comment by thesarfo** [comment]
@andrewtavis kindly feel free to reassign!🙂

**Comment by jamshaidali** [comment]
Hi @andrewtavis, this issue still seems to be open. I can take a look at it — I am currently working on setting it up locally. I will keep you posted. Could you please assign this issue to me?

**Comment by jamshaidali** [comment]
I was planning to take the same approach to add the throttle tests. Great to see that @knika01 as already raised the pull request. @knika01, I posted a suggestion on the PR — could you please take a look at it? In the meantime, @andrewtavis , could you please point me to some other areas where you would like me to improve the tests? I think we can still work on improving the test coverage for other parts of the code.

**Comment by andrewtavis** [comment]
Yes there definitely are some other areas where the tests can be improved. Do you want to run the coverage report as directed in the [backend section of the testing in the contributing guide](https://github.com/activist-org/activist/blob/main/CONTRIBUTING.md#backend-1) and make some suggestions for parts of the app that have low cover?

**Comment by andrewtavis** [comment]
Feel free to paste the output here and we can take a look/discuss together 😊

**Comment by andrewtavis** [comment]
I ended up running the coverage report for another issue, so the results are here (omitting all 100% lines):

```
Name                                       Stmts   Miss  Cover   Missing
------------------------------------------------------------------------
authentication/admin.py                       45     10    78%   64-69, 85-90
authentication/views.py                       90     13    86%   82-95, 187-197
communities/groups/serializers.py             54      5    91%   117, 137-142
communities/groups/views.py                  118     15    87%   67-68, 88-90, 112, 122-123, 137, 145-146, 174, 212, 236-237
communities/organizations/serializers.py      50      1    98%   86
communities/organizations/views.py           125     14    89%   64-65, 98-100, 144, 154-155, 189, 267, 308, 314, 338-339
content/serializers.py                        78     14    82%   86-91, 113-115, 148, 261-275
content/views.py                             163    111    32%   35-43, 49-60, 63-73, 76-86, 89-100, 103-112, 125-133, 139-150, 153-156, 159-170, 173-184, 187-196, 208-215, 221-244, 247-256, 259-270, 273-284, 287-296
core/settings.py                              51      2    96%   21-22
events/serializers.py                         57      8    86%   130, 138, 158-163, 197-199
events/views.py                              120     20    83%   54, 64-65, 85-87, 107, 116-117, 131, 138-139, 164, 171-172, 197, 203, 209, 233-234
utils/utils.py                                 9      3    67%   31, 54-58
------------------------------------------------------------------------
TOTAL                                       1555    216    86%
```

#1133 covers serializer testing, and #1132 is for views. From there we could look into `utils/utils.py`, finalize `core/settings.py` and also work on `authentication/admin.py`. Would you want to open up a PR for one/some of these, @jamshaidali? 😊

**Comment by jamshaidali** [comment]
Thanks for sharing the test reports, @andrewtavis. I would love to start working on improving the test coverage for the files you mentioned. The coverage for the utils and views files seems to be pretty low. However, it looks like we have the 80% threshold set for the minimum coverage. I think I can start by working on the utils.py test cases, raise their PR, and then look into the views.py coverage as well. Let me know if there is a designated issue you want me to be assigned to for these. I'll start working on them today.

**Comment by andrewtavis** [comment]
Thanks for all your efforts, @jamshaidali! You can link to this issue as that's where the discussion is. We can be a bit flexible on it, or if you'd prefer there to be a specific issue then you'd be welcome to open an issue for `utils.py` tests and then make it clear that you'll be working on it. I'll then assign in the morning 😊

**Comment by jamshaidali** [comment]
Sounds good @andrewtavis I have now created a new issue with the details to improve the test coverage for utils.py. Please take a look and assign it to me.

**Comment by andrewtavis** [comment]
Closed by #1271 :) Thanks for bringing this home, @to-sta, and also thanks @jamshaidali for your willingness to create and work on other issues after a bit of a confusing situation here with someone stepping in. Very much hope that we can work with you again soon! 😊
